Deprecate `PatternOptions` `storage` option

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | no
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE master BRANCH

- Are you adding documentation?
  - TARGET THE master BRANCH

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE master BRANCH

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE master BRANCH

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Since v2.10, the storage adapter have their own package. In v3.0 these, adapters have to register themselves to the `AdapterPluginManager` when the plugin manager is being instantiated and thus, this component wont be aware of which adapters are available without using a `PSR-11` container.
   
This commit aims to provide a proper dependency injection layer to all cache patterns while deprecating the old `storage` option for `PatternOptions`.


Closes #97

boesing committed:
qa: deprecate `storage` option for `PatternOptions`

Since v2.10, the storage adapter have their own package. In v3.0 these, adapters have to register themselves to the `AdapterPluginManager` when the plugin manager is being instantiated and thus, this component wont be aware of which adapters are available without using a `PSR-11` container.

This commit aims to provide a proper dependency injection layer to all cache patterns while deprecating the old `storage` option for `PatternOptions`.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented on src/Pattern/AbstractPattern.php:
```diff
@@ -8,37 +8,24 @@
         return $this;
     }
 
-    /**
```
These are part of the `PatternInterface`

boesing commented on src/Pattern/AbstractPattern.php:
```diff
@@ -8,37 +8,24 @@
-    /**
-     * Set pattern options
-     *
-     * @param  PatternOptions $options
```
These are part of the `PatternInterface`

boesing commented on src/Pattern/AbstractStorageCapablePattern.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    protected $storage;
+
+    public function __construct(?StorageInterface $storage = null, ?PatternOptions $options = null)
```
Will be changed to required arguments in v3.0

boesing commented:
@froschdesign I am not 100% sure but there are markdown issues with the tables. 

In the documentation HTML, it looks correct but using pipes within backticks while in a table column leads to display problems in non-HTML markdown editors.

So this is actually problematic:

```markdown
Option | Data Type | Default Value | Description
------ | --------- | ------------- | -----------
`storage` | `string | array | Laminas\Cache\Storage\StorageInterface` | none | Adapter used for reading and writing cached data.
```

https://gist.github.com/boesing/c37618ee975af0accbec2a6ce07cb458

It seems that this should be used instead (but leads to markdownlint issues due to the `code` usage):

```markdown
Option | Data Type | Default Value | Description
------ | --------- | ------------- | -----------
`storage` | <code>string &#124; array &#124; Laminas\Cache\Storage\StorageInterface</code> | none | Adapter used for reading and writing cached data.
```

https://gist.github.com/boesing/44e1ad9d9dd424d01b6e6393d50a4387

froschdesign commented:
We had the same problem in laminas-navigation: https://github.com/laminas/laminas-navigation/blob/2.12.x/docs/book/pages.md#common-page-options

Add a backslash:

```gfm
Option | Data Type | Default Value | Description
------ | --------- | ------------- | -----------
`storage` | `string\|array\|Laminas\Cache\Storage\StorageInterface` | none | Adapter used for reading and writing cached data.
```

Option | Data Type | Default Value | Description
------ | --------- | ------------- | -----------
`storage` | `string\|array\|Laminas\Cache\Storage\StorageInterface` | none | Adapter used for reading and writing cached data.

boesing committed:
qa: escape pipes in markdown to support more markdown reader

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

froschdesign commented on docs/book/pattern/callback-cache.md:
```diff
@@ -5,42 +5,40 @@ The callback cache pattern caches the results of arbitrary PHP callables.
-    'cache_output' => true,
-]);
+/** @var StorageInterface $storage */
+$storage = null; // Can be any instance of StorageInterface
```
The pattern section contains an introduction page, it would be good if a standard storage adapter for the documentation of the cache patterns is defined there. This can be done in blockquote and can then be linked.

For laminas-i18n I used the filesystem adapter: https://docs.laminas.dev/laminas-i18n/translator/caching/

boesing replied:
Hm, I've tried something but not sure if thats what you had in mind. Feel free to review.

boesing committed:
docs: add `Standard Storage Adapter` example to the pattern cache quick start section

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
@froschdesign If you have some time, I'd love to get feedback on my attempt to provide a `Standard Storage Adapter` documentation.

froschdesign commented on docs/book/pattern/callback-cache.md:
````diff
@@ -5,42 +5,50 @@ The callback cache pattern caches the results of arbitrary PHP callables.
+);
 ```
 
+> ### Storage Adapter
+>
+> The `$storage` adapter can be any adapter which implements the `StorageInterface`. Check out the [Pattern Quick Start](./intro.md#quick-start)-Section for a standard adapter which can be used here.
+
````
Unfortunately, no links can be set in code examples, so this looks good.

froschdesign commented on docs/book/pattern/capture-cache.md:
```diff
@@ -42,9 +46,36 @@ Option | Data Type | Default Value | Description
+    ])
+);
+
+// TODO
```
Instead of adding a todo comment in the documentation, a new issue report should be created. Or is this an old comment?

boesing replied:
Uhm, good question TBH. Have to double check 😅 

----

Yup, thats in there for years probably. I will remove it as I dont really know what is left to do here.

froschdesign commented on docs/book/pattern/intro.md:
````diff
@@ -22,39 +22,38 @@ It's also possible to use a single instance of
+Pattern objects can be created
 by instantiating one of the `Laminas\Cache\Pattern\*Cache` classes.
 
+> ### Standard Storage Adapter
+>
+> A standard storage adapter which can be used with the pattern caches might be the APCu adapter.
+>
+> ```php
+> use Laminas\Cache\Storage\Adapter\Apcu;
+> $storage = new Apcu();
+> ```
````
```suggestion
> ### Standard Storage Adapter for Documentation
>
> A cache adapter needs a storage adapter. To be able to follow the examples in the documentation, the [adapter for the filesystem](https://docs.laminas.dev/laminas-cache/storage/adapter/#filesystem-adapter) or the [BlackHole adapter](https://docs.laminas.dev/laminas-cache/storage/adapter/#blackhole-adapter) can be used, for example.
>
> ```php
> $storage = new Laminas\Cache\Storage\Adapter\Filesystem();
> // or
> $storage = new Laminas\Cache\Storage\Adapter\BlackHole();
> ```
```

froschdesign commented on docs/book/pattern/output-cache.md:
````diff
@@ -5,30 +5,54 @@ The `OutputCache` pattern caches output between calls to `start()` and `end()`.
+
+## Examples
+
+### Caching simple View Scripts
+
+```php
+use Laminas\Cache\Pattern\OutputCache;
+use Laminas\Cache\Pattern\PatternOptions;
+
+$outputCache = new OutputCache(
+    $storage,
+    new PatternOptions()
+);
+
+$outputCache->start('mySimpleViewScript');
+include '/path/to/view/script.phtml';
+$outputCache->end();
+```
````
Avoid words like "simple".

```suggestion
### Caching a View Scripts

```php
use Laminas\Cache\Pattern\OutputCache;
use Laminas\Cache\Pattern\PatternOptions;

$outputCache = new OutputCache(
    $storage,
    new PatternOptions()
);

$outputCache->start('myViewScript');
include '/path/to/view/script.phtml';
$outputCache->end();
```
```

boesing replied:
Fixed, most of the documentation already existed. Not quite sure why all of them have changes.

froschdesign commented on docs/book/pattern/class-cache.md:
````diff
@@ -8,34 +8,65 @@ class being cached, and caches static properties.
+
+### Caching of Import Feeds
+
+```php
+use Laminas\Cache\Pattern\ClassCache;
+use Laminas\Cache\Pattern\PatternOptions;
+
+$cachedFeedReader = new ClassCache(
+    $storage,
+    new PatternOptions([
+        'class' => \Laminas\Feed\Reader\Reader::class,
+        
+        // The feed reader doesn't output anything,
+        // so the output doesn't need to be caught and cached:
+        'cache_output' => false,
+    ])
+);
+
+$feed = $cachedFeedReader->call("import", array('http://www.planet-php.net/rdf/'));
+
+// OR
+$feed = $cachedFeedReader->import('http://www.planet-php.net/rdf/');
+```
````
```suggestion
```php
use Laminas\Cache\Pattern\ClassCache;
use Laminas\Cache\Pattern\PatternOptions;
use Laminas\Feed\Reader\Reader;

$cachedFeedReader = new ClassCache(
    $storage,
    new PatternOptions([
        'class' => Reader::class,
        
        // The feed reader doesn't output anything,
        // so the output doesn't need to be caught and cached:
        'cache_output' => false,
    ])
);

$feed = $cachedFeedReader->call("import", ['https://github.com/laminas/laminas-cache/releases.atom']);
// or
$feed = $cachedFeedReader->import('https://github.com/laminas/laminas-cache/releases.atom');
```
```

froschdesign commented:
@boesing 
Sorry for the _very_ late response. Please ping me, if you need help!

boesing committed:
qa: remove `TODO`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: optimize standard storage adapter documentation

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: update documentation header for the example of caching view scripts

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
docs: optimize cached feed reader documentation

- replaced dead links from example with github release notes
- imported laminas-feed `Reader` class

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented:
@froschdesign I'll merge this for now. We can optimize documentation in dedicated branches. I'd rather go moving forward with the release of v2 so we can finally kick out v3.
